osquery/events/darwin/openbsm.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <bsm/libbsm.h>
+
+#include <bsm/audit.h>
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+
+#include "openbsm.h"
+
+namespace osquery {
+
+/// The OpenBSM subsystem may have a performance impact on the system.
+FLAG(bool,
+     disable_openbsm,
+     true,
+     "Disable receiving events from the audit subsystem");
+
+REGISTER(OpenBSMEventPublisher, "event_publisher", "openbsm");
+
+Status OpenBSMEventPublisher::setUp() {
+  if (FLAGS_disable_openbsm) {
+    return Status(1, "Publisher disabled via configuration");
+  }
+  audit_pipe_ = fopen("/dev/auditpipe", "r");
+  if (audit_pipe_ == nullptr) {
+    LOG(WARNING) << "The auditpipe couldn't be opened.";
+    return Status(1, "Could not open OpenBSM pipe");
+  }
+  return Status(0);
+}
+
+void OpenBSMEventPublisher::configure() {}
+
+void OpenBSMEventPublisher::tearDown() {
+  if (audit_pipe_ != nullptr) {
+    fclose(audit_pipe_);
+  }
+}
+
+Status OpenBSMEventPublisher::run() {
+  if (audit_pipe_ == nullptr) {
+    return Status(1, "No open audit_pipe");
+  }
+  tokenstr_t tok;
+  auto reclen = 0;
+  auto bytesread = 0;
+  auto event_id = 0;
+  auto buffer = static_cast<unsigned char*>(nullptr);
+  std::vector<tokenstr_t> tokens{};
+
+  while ((reclen = au_read_rec(audit_pipe_, &buffer)) != -1) {
```

Comment:
Will this block if there's no data remaining? What happens if the daemon is shutting down and no events are available?